### PR TITLE
Issue #33 - Add password notes interface

### PR DESCRIPTION
**Pull request by kzar**
I've had a go at adding the interface required for password notes. Still a little rough and obviously it doesn't work since we can't store the notes yet. For now I've skipped adding a button to view / edit notes from the all passwords page.

### CODE REVIEW

**Comment by palant** [comment]
This looks good at first glance. Note that this will have to change anyway as we simply cannot keep adding icons for every action. After the next release (the one adding autofill for user names) we'll need to add some kind of overflow menu, and this is where this should go into.

**Comment by kzar** [comment]
@palant OK I've updated this now, sorry for the delay.

File: data/images/ios-compose-outline.svg [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+	 width="512px" height="512px" viewBox="0 0 512 512" enable-background="new 0 0 512 512" xml:space="preserve">
+<g>
+	<polygon points="368,416 80,416 80,144 304.117,144 320.117,128 64,128 64,432 384,432 384,191.883 368,207.883 	"/>
+	<polygon points="237.623,284.084 227.917,274.377 399.24,103.054 387.927,91.74 208,271.651 208,304 240.216,304 420.261,124.016 
+		408.947,112.73 	"/>
+	<path d="M444.213,80.312l-12.465-12.466C429.084,65.467,425.556,64,421.746,64c-3.812,0-7.304,1.468-9.929,3.85L399.666,80
+		l0.08,0.08l11.293,11.293l21.02,21.02l12.15-12.15c2.383-2.625,3.791-6.117,3.791-9.929C448,86.504,446.592,82.975,444.213,80.312z
+		"/>
+</g>
+</svg>
```
Comment (palant) [inline data/images/ios-compose-outline.svg]: I think that `ios-paper-outline.svg` will be a better match here.

File: locale/en-US.properties [hunk 2]
```diff
@@ -53,8 +54,9 @@ passwords-label = Passwords:
 password-menu = All actions
 to-document = Fill in
 to-clipboard = Copy to clipboard
-bump-revision = Generate new password for this user name
 show-qrcode = Show as QR code
+view-edit-notes = View / Edit notes
```
Comment (palant) [inline locale/en-US.properties]: This should be two strings: "Add notes" and "View notes". Otherwise it won't be obvious whether you already have notes for a password.

File: locale/en-US.properties [hunk 3]
```diff
@@ -10,6 +10,7 @@ no-characters-selected = At least one character set has to be selected.
 password-value-required = Please enter the password you used on this website.
 password-type-generated = Generated password
 password-type-legacy = Legacy password\nPassword encrypted and stored
+password-notes-stored = Contains notes which have been encrypted and stored
```
Comment (palant) [inline locale/en-US.properties]: How about making this shorter: "Notes encrypted and stored"

File: data/panel/panel.html [hunk 4]
```diff
@@ -97,14 +98,18 @@
...
-        </a>
         <a href="#" id="menu-show-qrcode" class="password-menu-entry">
           <span class="show-qrcode-link iconic-link"></span>
           <span data-l10n-id="show-qrcode"></span>
         </a>
+        <a href="#" id="menu-notes" class="password-menu-entry">
+          <span class="notes-link iconic-link"></span>
+          <span data-l10n-id="view-edit-notes"></span>
+        </a>
+        <a href="#" id="menu-bump-revision" class="password-menu-entry">
```
Comment (palant) [inline data/panel/panel.html]: Why did you move that menu entry? It's a fairly important one, that's why I have it further up.

As you probably know, I'm not a fan of code reordering making it harder to see what the review is about.

File: data/panel/panel.html [hunk 5]
```diff
@@ -189,6 +194,20 @@
       </div>
     </form>
 
+    <form id="notes" class="slide" data-default-element="notes-textarea">
+      <label data-l10n-id="site"></label>
+      <div id="notes-website-name"></div>
+      <label data-l10n-id="user-name"></label>
+      <div id="notes-user-name"></div>
+      <label class="block-start"  for="notes-textarea" data-l10n-id="password-notes"></label>
```
Comment (palant) [inline data/panel/panel.html]: This string appears to be missing.

Also a nit: extra space before `for`

File: data/panel/panel.html [hunk 6]
```diff
@@ -189,6 +194,20 @@
...
+      <textarea id="notes-textarea"></textarea>
+      <div class="error" hidden></div>
```
Comment (palant) [inline data/panel/panel.html]: I don't think we need validation for this field - any value fine.

File: data/panel/passwordList.js [hunk 7]
```diff
@@ -294,6 +298,21 @@ function copyToClipboard(password)
...
+  passwords.getPassword(state.site, password.name, password.revision)
+    .then(value => require("./qrcode").show(value))
+    .catch(showPasswordMessage);
+}
+
+function showNotes(password)
+{
+  setActivePanel("notes");
+
+  $("notes-website-name").textContent = state.site;
```
Comment (palant) [inline data/panel/passwordList.js]: I'd rather have this updated in `notes.js` on state changes, for consistency.

File: data/panel/passwordList.js [hunk 8]
```diff
@@ -294,6 +298,21 @@ function copyToClipboard(password)
...
+  $("notes-user-name").textContent = password.name;
```
Comment (palant) [inline data/panel/passwordList.js]: Please update state before making the panel visible. Also, it might be a good idea to have UI modification in a function exported from `notes.js` - this way all the logic related to notes stays there (similar to `showQRCode` above). The end result will look like this:

```
function showNotes(password)
{
  passwords.getNotes(state.site, password.name, password.revision)
    .then(notes => require("./notes").edit(password, notes))
    .catch(showPasswordMessage);
}
```

File: data/allpasswords/allpasswords.html [hunk 9]
```diff
@@ -64,6 +64,9 @@
         <div class="password-info legacy">
           <div class="password-type" data-l10n-id="password-type-legacy"></div>
         </div>
+        <div class="password-info notes">
+          <div data-l10n-id="password-notes-stored"></div>
```
Comment (palant) [inline data/allpasswords/allpasswords.html]: Please add some code making this visible - and be it a dummy.

File: data/panel/panel.html [hunk 10]
```diff
@@ -189,6 +194,20 @@
       </div>
     </form>
 
+    <form id="notes" class="slide" data-default-element="notes-textarea">
+      <label data-l10n-id="site"></label>
+      <div id="notes-website-name"></div>
+      <label data-l10n-id="user-name"></label>
+      <div id="notes-user-name"></div>
```
Comment (palant) [inline data/panel/panel.html]: I realized that this needs to display revision number as well - similar to how it is being displayed in the passwords list.

File: data/images/ios-compose-outline.svg (hunk 1 above)
Comment (kzar) [inline data/images/ios-compose-outline.svg]: Done.

File: locale/en-US.properties (hunk 3 above)
Comment (kzar) [inline locale/en-US.properties]: Done.

File: locale/en-US.properties (hunk 2 above)
Comment (kzar) [inline locale/en-US.properties]: Done.

File: data/panel/panel.html (hunk 4 above)
Comment (kzar) [inline data/panel/panel.html]: Well I figured this ordering was more logical. Actions near the top are viewing actions and the actions at the bottom make modifications. Since you can either view or edit your notes with this new action I figured it belonged in the middle.

File: data/panel/panel.html (hunk 5 above)
Comment (kzar) [inline data/panel/panel.html]: Done.

File: data/panel/panel.html (hunk 6 above)
Comment (kzar) [inline data/panel/panel.html]: Done.

File: data/panel/passwordList.js (hunk 7 above)
Comment (kzar) [inline data/panel/passwordList.js]: Done.

File: data/panel/passwordList.js (hunk 8 above)
Comment (kzar) [inline data/panel/passwordList.js]: Done.

File: data/panel/panel.html (hunk 10 above)
Comment (kzar) [inline data/panel/panel.html]: Done.

File: data/allpasswords/allpasswords.html (hunk 9 above)
Comment (kzar) [inline data/allpasswords/allpasswords.html]: Done.

**Comment by kzar** [comment]
@palant addressed your comments, can you see my replies or do I need to click something to publish them?

File: data/panel/main.js [hunk 11]
```diff
@@ -21,6 +21,7 @@ function init()
   require("./passwordList.js");
   require("./generatePassword.js");
   require("./legacyPassword.js");
+  require("./notes.js");
```
Comment (palant) [inline data/panel/main.js]: You don't need to require this explicitly any more, it is loaded when notes are being used.

File: data/panel/notes.js [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+setResetHandler("notes", setActivePanel.bind(null, "password-list"));
+
+function edit(password, notes)
+{
+  $("notes-user-name").textContent = password.name;
+  $("notes-password-revision").textContent = password.revision;
+  $("notes-website-name").textContent = state.site;
+
+  setActivePanel("notes");
+  $("notes-textarea").value = notes;
```
Comment (palant) [inline data/panel/notes.js]: Please move that statement above `setActivePanel()` call as well - first update the UI, then make it visible.

File: data/panel/panel.scss [hunk 13]
```diff
@@ -423,3 +428,8 @@ a:hover
 {
   align-self: center;
 }
+
+#notes-textarea
+{
+  min-height: 300px;
+}
```
Comment (palant) [inline data/panel/panel.scss]: No styles for the revision field?

File: locale/en-US.properties [hunk 14]
```diff
@@ -53,8 +54,10 @@ passwords-label = Passwords:
 password-menu = All actions
 to-document = Fill in
 to-clipboard = Copy to clipboard
-bump-revision = Generate new password for this user name
 show-qrcode = Show as QR code
+view-notes = View notes
+edit-notes = Edit notes
```
Comment (palant) [inline locale/en-US.properties]: No, it should be "Add notes" in order to indicate that there are no notes yet. This string isn't used right now, want to fix that?

File: data/panel/main.js (hunk 11 above)
Comment (kzar) [inline data/panel/main.js]: Done.

File: data/panel/notes.js (hunk 12 above)
Comment (kzar) [inline data/panel/notes.js]: Calling setActivePanel seemes to clear the textarea.

File: data/panel/panel.scss (hunk 13 above)
Comment (kzar) [inline data/panel/panel.scss]: What styles should it have? Just appending the revision number to the user name seems consistent with how it's done elsewhere. (Actually that makes me think, perhaps "user name (revision)" would be better, since otherwise people might confuse the revision number as being part of the user name.)

File: locale/en-US.properties (hunk 14 above)
Comment (kzar) [inline locale/en-US.properties]: Done.

File: data/panel/panel.scss (hunk 13 above)
Comment (palant) [inline data/panel/panel.scss]: See `.password-revision` - revision should be small text, with `#` before it and slightly above the usual baseline.

File: data/panel/notes.js (hunk 12 above)
Comment (palant) [inline data/panel/notes.js]: Yes, it resets the form. Well, never mind then.

File: data/panel/panel.scss (hunk 13 above)
Comment (kzar) [inline data/panel/panel.scss]: Ahh, I had some old messed up entries from the last time I was working on this. Just removed those and created some new ones and see what you mean.

**Comment by kzar** [comment]
@palant Updated again.

File: data/panel/panel.scss [hunk 15]
```diff
@@ -266,12 +266,14 @@ a:hover
...
-.password-revision::before
+.password-revision::before,
+#notes-password-revision::before
 {
   content: "#";
 }
 
-.password-revision
+.password-revision,
+#notes-password-revision
```
Comment (palant) [inline data/panel/panel.scss]: How about just adding `class="password-revision"` to the element instead?

Comment (kzar) [inline data/panel/panel.scss]: Good idea, done.

**Comment by palant** [comment]
This has been merged with https://github.com/palant/easypasswords/commit/eb6c6c72f422882d4af4c8d498e82d2c9f281c9c.
